internal/apimeta/mapper.go

```diff
@@ -0,0 +1,81 @@
+package apimeta
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+)
+
+// ResourceMapper allows callers to map resources to kinds.
+type ResourceMapper interface {
+	// NamespacedKindFor returns the priority mapping GVK for a namespaced GR.
+	// If the GR doesn't exist, or isn't namespaced, returns an error.
+	NamespacedKindFor(gr schema.GroupResource) (schema.GroupVersionKind, error)
+}
+
+// NotNamespacedError is returned if the mapper can find the resource, but it is not namespaced.
+type NotNamespacedError struct {
+	GroupResource schema.GroupResource
+}
+
+func (e *NotNamespacedError) Error() string {
+	return fmt.Sprintf("resource %v is not namespaced", e.GroupResource)
+}
+
+func IsNotNamespacedError(err error) bool {
+	if err == nil {
+		return false
+	}
+	switch err.(type) {
+	case *NotNamespacedError:
+		return true
+	default:
+		return false
+	}
+}
+
+func NewAPIMapper(c *rest.Config) (ResourceMapper, error) {
+	restMapper, err := apiutil.NewDynamicRESTMapper(c)
+	if err != nil {
+		return nil, err
+	}
+	return &APIMapper{
+		RestMapper: restMapper,
+	}, nil
+
+}
+
+// APIMapper allows callers to discover apiserver metadata at runtime
+type APIMapper struct {
+	RestMapper meta.RESTMapper
+}
+
+// NamespacedKindFor implements ResourceMapper.NamespacedKindFor by using a DynamicRESTMapper
+// to discover resource types at runtime.
+func (r *APIMapper) NamespacedKindFor(gr schema.GroupResource) (schema.GroupVersionKind, error) {
+	resource := gr.WithVersion("")
+	kinds, err := r.RestMapper.KindsFor(resource)
+	if err != nil {
+		return schema.GroupVersionKind{}, err
+	}
+	// Not sure if this can ever happen? RestMapper.KindsFor does not specify if no
+	// matches will error out or not, so just making sure we do not get issues.
+	if len(kinds) == 0 {
+		return schema.GroupVersionKind{}, &meta.NoResourceMatchError{PartialResource: resource}
+	}
+
+	// KindsFor returns the list of potential kinds in priority order
+	gvk := kinds[0]
+	mapping, err := r.RestMapper.RESTMapping(gvk.GroupKind())
+	if err != nil {
+		return schema.GroupVersionKind{}, err
```

Comment:
Maybe wrap the error, e.g. `"while attempting to determine if kind is namespaced: %w"`

Reply:
Done, or were you thinking of another form of wrapping. Please let know if I didn't do it correctly.

Reply:
lgtm